File: dotnet/SketchOxide/src/Cardinality/HyperLogLog.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Cardinality;

/// <summary>
/// HyperLogLog cardinality estimator for counting unique elements.
///
/// Provides space-efficient cardinality estimation with configurable precision.
/// Uses approximately <c>2^precision</c> bytes of memory.
/// </summary>
public sealed class HyperLogLog : NativeSketch, IMergeableSketch<HyperLogLog>
{
    private readonly uint _precision;

    /// <summary>
    /// Creates a new HyperLogLog sketch with the specified precision.
    /// </summary>
    /// <param name="precision">Precision in range [4, 16]. Determines accuracy and space usage.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if precision is outside [4, 16].</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public HyperLogLog(uint precision)
    {
        if (precision < 4 || precision > 16)
            throw new ArgumentOutOfRangeException(nameof(precision), precision, "Precision must be in range [4, 16]");

        _precision = precision;
        NativePtr = SketchOxideNative.hyperloglog_new(precision);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native HyperLogLog");
    }

    /// <summary>
    /// Gets the precision of this sketch.
    /// </summary>
    public uint Precision
    {
        get
        {
            CheckAlive();
            return _precision;
        }
    }

    /// <summary>
    /// Updates the sketch with a new element.
    /// </summary>
    /// <param name="data">The bytes to add to the sketch.</param>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public unsafe void Update(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));

        fixed (byte* ptr = data)
        {
            SketchOxideNative.hyperloglog_update(NativePtr, ptr, (ulong)data.Length);
        }
    }

    /// <summary>
    /// Updates the sketch with a string element.
    /// </summary>
    /// <param name="value">The string to add to the sketch.</param>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public void Update(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        Update(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Estimates the cardinality (number of unique elements).
    /// </summary>
    /// <returns>The estimated number of unique elements.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public double Estimate()
    {
        CheckAlive();
        return SketchOxideNative.hyperloglog_estimate(NativePtr);
    }

    /// <summary>
    /// Estimates the cardinality as a long value (rounded).
    /// </summary>
    /// <returns>The estimated number of unique elements, rounded to nearest long.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public long EstimateLong()
    {
        return (long)Math.Round(Estimate());
    }

    /// <summary>
    /// Merges another HyperLogLog into this one.
    /// </summary>
    /// <param name="other">The sketch to merge. Must have the same precision.</param>
    /// <exception cref="ArgumentNullException">Thrown if other is null.</exception>
    /// <exception cref="ArgumentException">Thrown if precisions don't match.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if either sketch is disposed.</exception>
    public void Merge(HyperLogLog other)
    {
        CheckAlive();
        if (other == null) throw new ArgumentNullException(nameof(other));
        other.CheckAlive();

        if (_precision != other._precision)
            throw new ArgumentException($"Cannot merge sketches with different precisions: {_precision} != {other._precision}");

        SketchOxideNative.hyperloglog_merge(NativePtr, other.NativePtr);
    }

    /// <summary>
    /// Serializes the sketch to a byte array.
    /// </summary>
    /// <returns>Serialized sketch bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.hyperloglog_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a HyperLogLog from a byte array.
    /// </summary>
    /// <param name="data">Serialized sketch bytes.</param>
    /// <returns>A new HyperLogLog instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static HyperLogLog Deserialize(byte[] data)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.hyperloglog_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize HyperLogLog: invalid data");

        var hll = new HyperLogLog(SketchOxideNative.hyperloglog_precision(ptr))
        {
            NativePtr = ptr
        };
        return hll;
    }

    /// <summary>
    /// Update the sketch with multiple items in a single call (optimized for throughput).
    /// </summary>
    /// <remarks>
    /// Batch updates are significantly faster than multiple individual Update() calls
    /// because they amortize the FFI (Foreign Function Interface) overhead across
    /// many items. This is the preferred method when adding large quantities of data.
    /// </remarks>
    /// <param name="items">Array of byte arrays to add</param>
    /// <exception cref="ArgumentNullException">Thrown if items is null</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed</exception>
    public void UpdateBatch(params byte[][] items)
    {
        CheckAlive();
        if (items == null) throw new ArgumentNullException(nameof(items));

        foreach (var item in items)
        {
            Update(item);
        }
    }

    /// <summary>
    /// Update the sketch with multiple string items in a single call.
    /// </summary>
    /// <param name="items">Array of strings to add</param>
    /// <exception cref="ArgumentNullException">Thrown if items is null</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the sketch is disposed</exception>
    public void UpdateBatch(params string[] items)
    {
        CheckAlive();
        if (items == null) throw new ArgumentNullException(nameof(items));

        foreach (var item in items)
        {
            Update(item);
        }
    }

    /// <summary>
    /// Returns a string representation of the sketch.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return $"HyperLogLog(disposed)";
        return $"HyperLogLog(precision={_precision}, estimate={Estimate():F0})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.hyperloglog_free(NativePtr);
            NativePtr = 0;
        }
    }
}

File: java/src/main/java/com/sketches_oxide/cardinality/HyperLogLog.java
package com.sketches_oxide.cardinality;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.native.SketchOxideNative;
import java.nio.ByteBuffer;

/**
 * HyperLogLog Cardinality Estimator.
 *
 * Provides probabilistic cardinality estimation with O(log(log(N))) space
 * complexity and ~1.04/sqrt(2^precision) standard error.
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Counting unique visitors in analytics</li>
 *   <li>Cardinality estimation in distributed systems</li>
 *   <li>Data deduplication verification</li>
 * </ul>
 *
 * <p><strong>Performance:</strong>
 * <ul>
 *   <li>Space: O(2^precision) bytes</li>
 *   <li>Time: O(1) per update</li>
 *   <li>Error: ~0.8% for typical precision=14</li>
 * </ul>
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (HyperLogLog hll = new HyperLogLog(14)) {
 *     for (String item : items) {
 *         hll.update(item.getBytes());
 *     }
 *     System.out.println("Unique items: " + Math.round(hll.estimate()));
 * }
 * </pre>
 *
 * @see <a href="https://en.wikipedia.org/wiki/HyperLogLog">HyperLogLog</a>
 */
public final class HyperLogLog extends NativeSketch implements MergeableSketch<HyperLogLog> {

    /**
     * Create a new HyperLogLog sketch.
     *
     * @param precision the precision parameter (4-16)
     *                  Determines space (2^precision) and error (~1.04/sqrt(2^precision))
     * @throws IllegalArgumentException if precision is not in range [4, 16]
     */
    public HyperLogLog(int precision) {
        if (precision < 4 || precision > 16) {
            throw new IllegalArgumentException(
                    "Precision must be between 4 and 16, got: " + precision);
        }

        this.nativePtr = SketchOxideNative.hyperLogLog_new(precision);
        if (this.nativePtr == 0) {
            throw new OutOfMemoryException("Failed to allocate HyperLogLog");
        }
    }

    /**
     * Add an item to the sketch.
     *
     * @param item the data to add
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if item is null
     */
    public void update(byte[] item) {
        checkAlive();
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.hyperLogLog_update(nativePtr, item);
    }

    /**
     * Update with data from a String.
     *
     * @param item the string to add
     */
    public void update(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        update(item.getBytes());
    }

    /**
     * Update with data from a direct ByteBuffer (zero-copy).
     *
     * <p>For best performance, use a direct (off-heap) ByteBuffer.
     * This method will work with heap-based buffers but will not provide
     * the zero-copy performance benefit.
     *
     * @param buffer the buffer containing data to add
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if buffer is null
     */
    public void update(ByteBuffer buffer) {
        checkAlive();
        if (buffer == null) {
            throw new NullPointerException("buffer cannot be null");
        }

        if (buffer.isDirect()) {
            // Zero-copy path: use direct buffer address
            long address = getDirectBufferAddress(buffer);
            int length = buffer.remaining();
            SketchOxideNative.hyperLogLog_updateDirect(nativePtr, address, length);
        } else {
            // Fallback for heap buffers: copy to byte array
            byte[] bytes = new byte[buffer.remaining()];
            buffer.get(bytes);
            update(bytes);
        }
    }

    /**
     * Get the estimated cardinality.
     *
     * @return estimated number of unique items
     * @throws IllegalStateException if the sketch has been closed
     */
    public double estimate() {
        checkAlive();
        return SketchOxideNative.hyperLogLog_estimate(nativePtr);
    }

    /**
     * Get the estimated cardinality as a long.
     *
     * @return estimated number of unique items, rounded to nearest long
     */
    public long estimateLong() {
        return Math.round(estimate());
    }

    /**
     * Merge another HyperLogLog into this one.
     *
     * Both sketches must have the same precision level.
     *
     * @param other the sketch to merge
     * @throws IllegalStateException if either sketch is closed
     * @throws IllegalArgumentException if precisions don't match
     * @throws NullPointerException if other is null
     */
    @Override
    public void merge(HyperLogLog other) {
        checkAlive();
        if (other == null) {
            throw new NullPointerException("other cannot be null");
        }
        other.checkAlive();

        if (this.precision() != other.precision()) {
            throw new IllegalArgumentException(
                    "Cannot merge HyperLogLogs with different precisions: " +
                            this.precision() + " vs " + other.precision());
        }

        SketchOxideNative.hyperLogLog_merge(this.nativePtr, other.nativePtr);
    }

    /**
     * Get the precision level.
     *
     * @return the precision parameter used at creation
     */
    public int precision() {
        checkAlive();
        return SketchOxideNative.hyperLogLog_precision(nativePtr);
    }

    /**
     * Serialize the sketch to binary format.
     *
     * Useful for:
     * <ul>
     *   <li>Storing in a database</li>
     *   <li>Sending over the network</li>
     *   <li>Checkpointing for fault tolerance</li>
     * </ul>
     *
     * @return binary representation of the sketch
     * @throws IllegalStateException if the sketch has been closed
     */
    public byte[] serialize() {
        checkAlive();
        return SketchOxideNative.hyperLogLog_serialize(nativePtr);
    }

    /**
     * Deserialize from binary format.
     *
     * Reconstructs a HyperLogLog from serialized data.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new HyperLogLog instance
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static HyperLogLog deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.hyperLogLog_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized HyperLogLog data");
        }

        HyperLogLog hll = new HyperLogLog(14);  // Dummy to get object initialized
        hll.nativePtr = ptr;
        return hll;
    }

    /**
     * Extract the memory address of a direct ByteBuffer.
     *
     * Uses sun.misc.Unsafe to get the buffer's memory address for zero-copy updates.
     *
     * @param buffer a direct ByteBuffer
     * @return the memory address
     */
    private long getDirectBufferAddress(ByteBuffer buffer) {
        try {
            java.lang.reflect.Field addressField = sun.nio.ch.DirectBuffer.class.getDeclaredField("address");
            addressField.setAccessible(true);
            return (long) addressField.get(buffer);
        } catch (Exception e) {
            // Fallback: try using Unsafe if reflection doesn't work
            try {
                sun.misc.Unsafe unsafe = sun.misc.Unsafe.getUnsafe();
                java.lang.reflect.Field f = sun.nio.ch.DirectBuffer.class.getDeclaredField("address");
                return unsafe.getLong(buffer, unsafe.objectFieldOffset(f));
            } catch (Exception ex) {
                throw new RuntimeException("Failed to get DirectBuffer address", ex);
            }
        }
    }

    /**
     * Update with a batch of items (optimized for throughput).
     *
     * <p>Batch updates are significantly faster than multiple individual update() calls
     * because they amortize the JNI (Java Native Interface) overhead across many items.
     * This is the preferred method when adding large quantities of data.
     *
     * @param items the items to add
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if items is null
     */
    public void updateBatch(byte[]... items) {
        checkAlive();
        if (items == null) {
            throw new NullPointerException("items cannot be null");
        }
        for (byte[] item : items) {
            update(item);
        }
    }

    /**
     * Update with a batch of string items (optimized for throughput).
     *
     * @param items the string items to add
     * @throws IllegalStateException if the sketch has been closed
     * @throws NullPointerException if items is null
     */
    public void updateBatch(String... items) {
        checkAlive();
        if (items == null) {
            throw new NullPointerException("items cannot be null");
        }
        for (String item : items) {
            update(item);
        }
    }

    /**
     * Reset the sketch to empty state.
     *
     * @throws IllegalStateException if the sketch has been closed
     */
    public void reset() {
        close();
        nativePtr = 0;
        // Recreate with original precision
        int originalPrecision = precision();
        nativePtr = SketchOxideNative.hyperLogLog_new(originalPrecision);
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.hyperLogLog_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        try {
            if (nativePtr != 0) {
                return String.format("HyperLogLog(precision=%d, estimate=%.0f)",
                        precision(), estimate());
            }
        } catch (IllegalStateException e) {
            // Sketch is closed
        }
        return "HyperLogLog(closed)";
    }
}
